File: facade/src/main/java/org/tdf/sunflower/types/StorageWrapper.java
package org.tdf.sunflower.types;

import lombok.RequiredArgsConstructor;
import org.tdf.common.store.Store;
import org.tdf.common.util.HexBytes;
import org.tdf.common.util.RLPUtil;
import org.tdf.rlpstream.Rlp;
import org.tdf.rlpstream.RlpList;

import java.util.*;

@RequiredArgsConstructor
public class StorageWrapper {
    private final HexBytes prefix;
    private final Store<HexBytes, HexBytes> store;


    public StorageWrapper(Store<HexBytes, HexBytes> store) {
        this(HexBytes.empty(), store);
    }

    private HexBytes verifyAndPrefix(HexBytes key) {
        return prefix.concat(key);
    }

    public Set<HexBytes> keySet() {
        if (this.prefix.isEmpty())
            throw new UnsupportedOperationException();
        HexBytes keys = store.get(prefix);
        return keys == null || keys.size() == 0 ?
            new TreeSet<>() :
            new TreeSet<>(
                Arrays.asList(RLPUtil.decode(keys, HexBytes[].class))
            );
    }


    private void addKey(HexBytes key) {
        Set<HexBytes> keySet = keySet();
        keySet.add(key);
        store.set(prefix, HexBytes.fromBytes(Rlp.encode(keySet)));
    }

    public void save(HexBytes key, Object o) {
        HexBytes prefixed = verifyAndPrefix(key);
        // if this is prefix store, add key
        if (!this.prefix.isEmpty()) {
            addKey(key);
        }
        store.set(prefixed, RLPUtil.encode(o));
    }

    private void removeKey(HexBytes key) {
        Set<HexBytes> keySet = keySet();
        keySet.remove(key);
        store.set(prefix, HexBytes.fromBytes(Rlp.encode(keySet)));
    }

    public void remove(HexBytes key) {
        HexBytes prefixed = verifyAndPrefix(key);
        // if this is prefix store, add key
        if (!this.prefix.isEmpty()) {
            removeKey(key);
        }
        store.remove(prefixed);
    }

    public <T> T get(HexBytes key, Class<T> clazz, T defaultValue) {
        HexBytes prefixed = verifyAndPrefix(key);
        HexBytes h = store.get(prefixed);
        if (h == null)
            return defaultValue;
        return RLPUtil.decode(h, clazz);
    }

    public <T> List<T> getList(HexBytes key, Class<T> clazz, List<T> defaultValue) {
        HexBytes prefixed = verifyAndPrefix(key);
        HexBytes h = store.get(prefixed);
        if (h == null)
            return defaultValue;
        List<T> ret = new ArrayList<>();
        RlpList li = Rlp.decodeList(h.getBytes());

        for(int i = 0; i < li.size(); i++) {
            ret.add(
                Rlp.decode(li.rawAt(i), clazz)
            );
        }
        return ret;
    }

    public <T extends Comparable<T>> Set<T> getSet(HexBytes key, Class<T> clazz, Set<T> defaultValue) {
        HexBytes prefixed = verifyAndPrefix(key);
        HexBytes h = store.get(prefixed);
        if (h == null)
            return defaultValue;
        TreeSet<T> ret = new TreeSet<>();

        RlpList li = Rlp.decodeList(h.getBytes());

        for(int i = 0; i < li.size(); i++) {
            ret.add(
                Rlp.decode(li.rawAt(i), clazz)
            );
        }
        return ret;
    }
}

File: facade/src/main/java/org/tdf/sunflower/types/StorageWrapper.kt
package org.tdf.sunflower.types

import com.github.salpadding.rlpstream.Rlp
import org.tdf.common.store.Store
import org.tdf.common.util.*
import java.util.*

class StorageWrapper(private val prefix: HexBytes, private val store: Store<HexBytes, HexBytes>) {
    constructor(store: Store<HexBytes, HexBytes>) : this(HexBytes.empty(), store)

    private fun verifyAndPrefix(key: HexBytes): HexBytes {
        return prefix + key
    }

    private fun keySet(): MutableSet<HexBytes> {
        if (prefix.isEmpty()) throw UnsupportedOperationException()
        val keys = store[prefix]
        return if (keys == null || keys.isEmpty()) TreeSet() else
            TreeSet(keys.decode(Array<HexBytes>::class.java).toList())
    }

    private fun addKey(key: HexBytes) {
        val keySet = keySet()
        keySet.add(key)
        store[prefix] = keySet.rlp().hex()
    }

    fun save(key: HexBytes, o: Any) {
        val prefixed = verifyAndPrefix(key)
        // if this is prefix store, add key
        if (!prefix.isEmpty()) {
            addKey(key)
        }
        store[prefixed] = o.rlp().hex()
    }

    private fun removeKey(key: HexBytes) {
        val keySet = keySet()
        keySet.remove(key)
        store[prefix] = keySet.rlp().hex()
    }

    fun remove(key: HexBytes) {
        val prefixed = verifyAndPrefix(key)
        // if this is prefix store, add key
        if (!prefix.isEmpty()) {
            removeKey(key)
        }
        store.remove(prefixed)
    }

    operator fun <T> get(key: HexBytes, clazz: Class<T>, defaultValue: T): T {
        val prefixed = verifyAndPrefix(key)
        val h = store[prefixed] ?: return defaultValue
        return h.decode(clazz)
    }

    fun <T> getList(key: HexBytes, clazz: Class<T>, defaultValue: MutableList<T>?): MutableList<T>? {
        val prefixed = verifyAndPrefix(key)
        val h = store[prefixed]
        if (h == null || h.isEmpty()) return defaultValue
        val ret: MutableList<T> = mutableListOf()
        val li = Rlp.decodeList(h.bytes)
        for (i in 0 until li.size()) {
            ret.add(li.rawAt(i).decode(clazz))
        }
        return ret
    }

    fun <T : Comparable<T>> getSet(key: HexBytes, clazz: Class<T>, defaultValue: MutableSet<T>?): MutableSet<T>? {
        val prefixed = verifyAndPrefix(key)
        val h = store[prefixed]
        if (h == null || h.isEmpty()) return defaultValue
        val ret = TreeSet<T>()
        val li = Rlp.decodeList(h.bytes)
        for (i in 0 until li.size()) {
            ret.add(li.rawAt(i).decode(clazz))
        }
        return ret
    }
}